Jenkinsfile

```diff
@@ -636,38 +642,70 @@ def doLinuxCrossCompile(String target, String buildType, String emulator = null)
                             -DCPACK_SYSTEM_NAME=Linux-${target} \
                             ..
                     """
 
                     runAndCollectWarnings(script: "ninja")
 
-                    if (emulator != null) {
+                    if (testOptions != null) {
+                        stash includes: 'test/**/*', name: "realm-tests-Linux-${target}"
+                    } else {
+                        sh 'cpack'
+                        archiveArtifacts '*.tar.gz'
+                        def stashName = "linux-${target}___${buildType}"
+                        stash includes:"*.tar.gz", name:stashName
+                        publishingStashes << stashName
+                    }
+                }
+            }
+
+            if (testOptions != null) {
+                def runTests = { emulated ->
+                    dir('build-dir') {
+                        unstash "realm-tests-Linux-${target}"
+                        def runner = emulated ? testOptions.emulator : ''
                         try {
                             def environment = environment()
                             environment << 'UNITTEST_PROGRESS=1'
-                            environment << 'UNITTEST_FILTER=- Thread_RobustMutex*'  // robust mutexes can't work under qemu
+                            if (emulated) {
+                                environment << 'UNITTEST_FILTER=- Thread_RobustMutex*'  // robust mutexes can't work under qemu
+                            }
                             withEnv(environment) {
                                 sh """
                                     cd test
                                     ulimit -s 256 # launching thousands of threads in 32-bit address space requires smaller stacks
-                                    LD_LIBRARY_PATH=/usr/arm-linux-gnueabihf/lib ${emulator} realm-tests
+                                    ${runner} ./realm-tests
                                 """
                             }
                         } finally {
                             dir('..') {
-                                recordTests("Linux-${target}-${buildType}")
+                                def suffix = emulated ? '-emulated' : ''
+                                recordTests("Linux-${target}-${buildType}${suffix}")
                             }
                         }
-                    } else {
-                        sh 'cpack'
-                            archiveArtifacts '*.tar.gz'
-                            def stashName = "linux-${target}___${buildType}"
-                            stash includes:"*.tar.gz", name:stashName
-                            publishingStashes << stashName
                     }
                 }
+
+                // Note: Jenkins BlueOcean view doesn't display nested parallelism correctly while it is running.
+                // It waits until all tasks have finished before showing subnodes.
```

Comment:
I could also remove the parallelism. It seems worth it in general if it will let tests finish faster. On the other hand, if it makes CI more flaky, maybe not?

Reply:
Jenkins doesn't like it too much when there are nested parallel pipelines. How about if you refactor this function to either run a test on the native node, or with qemu, and just add two branches to the top parallel build? Sure, we'll end up compiling the exact same way twice, but we've got plenty of generic build nodes for that.